flang/include/flang/Lower/Support/BoxValue.h

```diff
@@ -269,18 +230,18 @@ class AbstractIrBox : public AbstractBox {
 };
 
 /// An entity described by a fir.box value that cannot be read into
-/// another BoxValue category, either because the fir.box may be an
+/// another ExtendedValue category, either because the fir.box may be an
 /// absent optional and we need to wait until the user is referencing it
 /// to read it, or because it contains important information that cannot
 /// be exposed in FIR (e.g. non contiguous byte stride).
 /// It may also store explicit bounds or length parameters that were specified
 /// for the entity.
-class IrBoxValue : public AbstractIrBox {
+class BoxValue : public AbstractIrBox {
```

Comment:
Shouldn't this inherit the `AbstractArrayBox` interface for consistency? It looks like the only difference is the minor renaming of `extents` to `explicitExtents`.

For the boxing of CHARACTER (e.g. CHARACTER, POINTER), is the intention to add the LEN to `explicitParams`?

Reply:
I am not sure, there is code out there that expect that `AbstractArrayBox` extents are always filled-in, which is not the case for this one. For instance, `AbstractArrayBox` has a `rank` method that returns the extent size and would just be wrong here. Plus I feel like inheriting `AbstractArrayBox` signals that this is a class is reserved to arrays.

> For the boxing of CHARACTER (e.g. CHARACTER, POINTER), is the intention to add the LEN to `explicitParams` ?

When there is an explicit one, that is what is happening yes, otherwise it is in the `fir.box`.

Reply:
It was not intended to imply that a boxed value (which can be anything) must be an array. The point was to make all the extended values that _may_ be arrays have identical interfaces, instead of interface du jour.

Reply:
Ok, I tried to make `BoxValue` inherit `AbstractArrayBox`, but I am getting compilation errors because they are conflicts between methods inherited form  `AbstractArrayBox` and  `AbstractIrBox` (for instance the `rank()` method).

Reply:
The changes here https://github.com/flang-compiler/f18-llvm-project/commit/29f66052187c8fdbe296e773b75d17dab246bc7a work for me.

Reply:
Thanks for making this work, I had not thought of making `AbstractIrBox` inherits from `AbstractArrayBox`. That looks reasonable to me then. I used your patch and updated the comment /cleaned-up the old members.